sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java

```diff
@@ -119,27 +124,28 @@ private MethodHandle locateResponseConstructor(Class<?> responseClass) {
         final int paramCount = handle.type().parameterCount();
         switch (paramCount) {
             case 3:
-                return constructResponse(handle, THREE_PARAM_ERROR, LOGGER, httpRequest, responseStatusCode,
+                return constructResponse(handle, THREE_PARAM_ERROR, httpRequest, responseStatusCode,
                     responseHeaders);
             case 4:
-                return constructResponse(handle, FOUR_PARAM_ERROR, LOGGER, httpRequest, responseStatusCode,
+                return constructResponse(handle, FOUR_PARAM_ERROR, httpRequest, responseStatusCode,
                     responseHeaders, bodyAsObject);
             case 5:
-                return constructResponse(handle, FIVE_PARAM_ERROR, LOGGER, httpRequest, responseStatusCode,
+                return constructResponse(handle, FIVE_PARAM_ERROR, httpRequest, responseStatusCode,
                     responseHeaders, bodyAsObject, decodedResponse.getDecodedHeaders());
             default:
-                return monoError(LOGGER, new IllegalStateException(INVALID_PARAM_COUNT));
+                throw LOGGER.logExceptionAsError(new IllegalStateException(INVALID_PARAM_COUNT));
         }
     }
 
-    private static Mono<Response<?>> constructResponse(MethodHandle handle, String exceptionMessage,
-        ClientLogger logger, Object... params) {
-        return Mono.defer(() -> {
-            try {
-                return Mono.just((Response<?>) handle.invokeWithArguments(params));
-            } catch (Throwable throwable) {
-                return monoError(logger, new RuntimeException(exceptionMessage, throwable));
+    private static Response<?> constructResponse(MethodHandle handle, String exceptionMessage, Object... params) {
+        try {
+            return (Response<?>) handle.invokeWithArguments(params);
+        } catch (Throwable throwable) {
+            if (throwable instanceof Error) {
+                throw (Error) throwable;
             }
```

Comment:
if this is a pattern we may invent some util in the core for this. (and add checkstyle rule to enforce usage)